### ISSUE TITLE
A way to get instance features closest to given image?

### ISSUE THREAD

**Issue by alex4727** [facebookresearch/ic_gan#26]
Hi, 
I am trying to regenerate an image using icgan (stylegan backbone, coco).
For projected w vector, I use codes from projector.py and the problem is about instance features. 
To regenerate using icgan, I need to obtain instance features closest to given image (image to regenerate).
I'll probably need to get a feature of that image and find the closest cluster index among pre-trained 1000 k-means. Are there any  code segments or functions to do this simply..? I do see some related codes in ILSVRC_HDF5_feats class but is there anyway to do this simply without creating a that big dataset class?

**Comment by ArantxaCasanova** [comment]
Hello, 
you could use most of the code already present in the [Colab notebook](https://github.com/facebookresearch/ic_gan/blob/main/inference/icgan_colab.ipynb):
you could try to 1) first load the feature extractor of choice (function `load_feature_extractor`),  2) pre-process the image of your choice by taking a central crop, resizing the image and normalizing by the ImageNet stats (function `preprocess_input_image`) and 3) obtain the feature vector to later find it's closest instance features out of these 1000 pre-saved ones from ImageNet:
```
    out_ = preprocess_generated_image(input_image)
    with torch.no_grad():
      out_features, _ = feature_extractor(out_.cuda())
    out_features/=torch.linalg.norm(out_features,dim=-1, keepdims=True)
    dists = sklearn.metrics.pairwise_distances(
            out_features.cpu(), reference_features.cpu(), metric="euclidean", n_jobs=-1)
```
where `reference_features` will be the loaded instance feature vectors, that will be loaded and divided by their norm. 


I hope this helps.

**Comment by alex4727** [comment]
Thanks, that was just what I needed..! 
One more question,, are there any reasons for using 1000 clusters with coco? Coco dataset contains 182 classes, so I wondered if the results would get better with clusters made with 182 centers. 
I'd to know if there are any experiments about number of centers in coco dataset

**Comment by ArantxaCasanova** [comment]
No, we have not extensively analyzed the clustering in COCO, but in our experimentation 1000 clusters worked well for COCO too. You are more than welcome to submit some additional information/insights in this thread if you end up exploring with this!

**Comment by yaohwang** [comment]
hi, how to generate stored_instances.tar.gz with my own dataset(only images).
with tutorial to prepare data I only get son hdf5 file and some other npz files, 
without stored_instances/%s_res%i_rn50_%s_kmeans_k1000_instance_features.npy needed by generate_images.py

**Comment by ArantxaCasanova** [comment]
Hi, @yaohwang, please note that the topic of this issue is a bit different from the topic of your question.
You can refer to this answer: https://github.com/facebookresearch/ic_gan/issues/24#issuecomment-953152475 to see what you need to do in order to create the file needed by generate_images.py. 
If you have follow up questions, please open a new issue.
